### PR TITLE
Data science

### PR DESCRIPTION
**Pull request by elialbert**
fixed the formatting as requested by @zsyed1

### CODE REVIEW

File: data_science/data_loader_spec.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+require_relative 'data_science'
+require_relative 'data_loader'
+
+describe DataLoader do
+  describe "optimizer loading" do
+    it "loads data" do
+      optimizer = Optimizer.new [
+        { cohort: "A", result: 1 },
+        { cohort: "A", result: 1 },
+      ]
```
Comment (zsyed1) [inline data_science/data_loader_spec.rb]: Lines 7-10: You can move this into a `let` statement:
```
let (:optimizer) do
  Optimizer.new [
    ...
    ...
  ]
end
```

File: data_science/data_loader_spec.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      expect(optimizer.data[:A][:successes]).to eq(2)
+    end
+
+    it "loads from filename" do
+      optimizer = Optimizer.new "data_export_2014_06_20_15_59_02.json"
```
Comment (zsyed1) [inline data_science/data_loader_spec.rb]: same thing - move this into a `let. You can distinguish the names by `let(:simple_data)` for the one above.

File: data_science/data_science.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+  def initialize(initial_data)
+    @data = DataLoader.load_data(initial_data)
+  end
+
+  def simple_counts
+    [:A, :B].each_with_object({}) do |cohort, result|
+      result[cohort] = { sample_size: @data[cohort].values.inject(:+),
+                         conversions: @data[cohort][:successes] }
+    end
```
Comment (zsyed1) [inline data_science/data_science.rb]: Lines 13-16:
Probably want to move the per-cohort logic into a separate method:
```
def simple_counts
  { A: simple_count_for(:A), B: simple_count_for(:B) }
end

def simple_count_for(cohort)
  data = @data[cohort]
  { sample_size: data.values.inject(:+), conversions: data[:successes] }
end
```

File: data_science/data_science.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  end
+
+  def conversion_rates
+    counts = simple_counts
+    [:A, :B].each_with_object({}) do |cohort, result|
+      result[cohort] = ABAnalyzer.confidence_interval(
+        counts[cohort][:conversions], counts[cohort][:sample_size], 0.95)
+    end
```
Comment (zsyed1) [inline data_science/data_science.rb]: Same suggestion as above

File: data_science/optimizer_spec.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
+require_relative 'data_science'
+
+describe Optimizer do
+  before :each do
+    @optimizer = Optimizer.new 'test_data.json'
```
Comment (zsyed1) [inline data_science/optimizer_spec.rb]: Use `let` instead. That's our convention :smile:

File: data_science/optimizer_spec.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    it "counts number of conversions per cohort" do
+      expect(@optimizer.simple_counts[:A][:conversions]).to eq(12)
+    end
+  end
+
+  describe "calculations" do
+    it "should calculate conversion rates" do
+      result = @optimizer.conversion_rates
+      expect(result[:A].class).to eq(Array)
```
Comment (zsyed1) [inline data_science/optimizer_spec.rb]: Not sure if you need this.

File: data_science/optimizer_spec.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      expect(result[:A][0]).to be < result[:A][1]
+    end
+
+    it "should do chisquared calculations" do
+      expect(@optimizer.result_confidence).to be < 1
```
Comment (zsyed1) [inline data_science/optimizer_spec.rb]: You really need to check that the chi-square or p-value is correct. I forget the numbers - like x=3.63 and p=0.96

File: data_science/test_data.json [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  {"cohort": "B", "result": 0},
+  {"cohort": "A", "result": 1}, 
+  {"cohort": "A", "result": 0}, 
+  {"cohort": "B", "result": 0},
+  {"cohort": "A", "result": 1}, 
+  {"cohort": "A", "result": 0}, 
+  {"cohort": "B", "result": 1},
+  {"cohort": "A", "result": 0}, 
+  {"cohort": "A", "result": 0}, 
+  {"cohort": "B", "result": 0}]
```
Comment (zsyed1) [inline data_science/test_data.json]: Nice work! Very concise. I would recommend making the changes and resubmitting for a quick review.

**Comment by paulghaddad** [comment]
@elialbert Nice work! As Zamir suggested, please make his recommendations and resubmit.

**Comment by elialbert** [comment]
will do

**Pull request closed without merging**